### DISCUSSION TITLE
vehicle delivers late

### DISCUSSION THREAD

**Discussion by kazitanvirahsan** [google/or-tools#2851] [Routing (and legacy CP) questions]
Hai Everyone,

Say in a scenario, we have 4 loads and their pickups and deliveries are (P1, P2, P3, P4) and (D1, D2, D3, D4) respectively.

All deliveries (D1,D2,D3,D4) are to the same location (L1).

We noticed that in some solutions, the vehicle picks up loads P1, P2, P3 (max vehicle capacity) then delivers D1, D2 then it picks P4, and delivers D3,D4.

P1 -> P2 -> P3 -> (D1 -> D2)  -> P4 -> (D3 ->D4)
................................t1(L1)..........................t2(L1) 
(t)------------------------------------------------------------------> 

Here P3 gets delivered at t2 in L1.
 

Is there any way we can force the tool to deliver P3 at t1 in L1? So instead of the above can, can the tool be able to make the visit as below?

P1 -> P2 -> P3 -> (D1 -> D2 -> D3) -> P4 -> (D4)
......................................t1(L1).........................t2(L1) 

I look forward to your help. 

Kazi

**Comment by Mizux** [comment]
see other thread with gate node so you can have:

* Create a dimension "L1_load" with +1 for `P*` and -1 for any `D*`
* add constraints,  for all i: `l1_load.CumulVar(Gi) == 0` aka when leaving L1 i.e.
 visiting any node `Gi` vehicle must not carry any L1 goods.
* for all i: `NextVar(Di) := D* + G_*` aka force vehicle to visit a `Gi` when leaving L1

Basically you want:
```
P1 -> P2 -> ((D1 -> D2) -> G_1) -> P3 -> P4 -> ((D3 ->D4) -> G_2)
```
